### PR TITLE
Hashtable iteration

### PR DESCRIPTION
**Pull request by johnfxgalea**
Added a new function for dynamorio's hashtable implementation. 

The function allows the calling of a function for each payload entered in the hashtable. It is useful if updates need to be performed for all payloads. 

Abstracting this functionality from the implementation of the hashtable would typically require managing an external data-structure, such as a list, that contains the payloads or the keys (where payloads are then retrieved via look-up operations). The proposed approach avoids this.

### CODE REVIEW

File: ext/drcontainers/hashtable.h [hunk 1]
```diff
@@ -193,6 +193,13 @@ hashtable_remove(hashtable_t *table, void *key);
 bool
 hashtable_remove_range(hashtable_t *table, void *start, void *end);
 
+/**
+ * Calls the apply_func for each payload
+ */
+void
+hashtable_apply_to_all_payloads(hashtable_t *table, void (*apply_func)(void *payload));
+
+
```
Comment (fhahn) [inline ext/drcontainers/hashtable.h]: Only one new line here

File: ext/drcontainers/hashtable.c [hunk 2]
```diff
@@ -457,6 +457,23 @@ hashtable_remove_range(hashtable_t *table, void *start, void *end)
     return res;
 }
 
+
+void
+hashtable_apply_to_all_payloads(hashtable_t *table, void (*apply_func)(void *payload))
+{
+    if (apply_func){
```
Comment (fhahn) [inline ext/drcontainers/hashtable.c]: I would make apply_func required, i.e. replacing the if with an assert.

File: ext/drcontainers/hashtable.h (hunk 1 above)
Comment (fhahn) [reply ext/drcontainers/hashtable.h]: Could you add doxygen comments for the parameters?

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by derekbruening** [comment]
Echoing @fhahn, thank you for submitting a new feature!

If we're adding an iterator interface, it seems best to look at the existing iterations and ensure the interface meets their needs, and IMHO this CL (or maybe a follow-up, but we should make sure now that we won't have to change the interface) should replace their raw iterations with the new interface.  WDYT?

There are uses in several tools, libraries, and tests:

```
$ git grep HASHTABLE_SIZE | egrep -v '^core|drcontainers/hash'
clients/drcachesim/tracer/raw2trace.cpp:    for (uint i = 0; i < HASHTABLE_SIZE(decode_cache.table_bits); i++) {
clients/drcov/postprocess/drcov2lcov.cpp:    for (i = 0; i < HASHTABLE_SIZE(module_htable.table_bits); i++) {
clients/drcov/postprocess/drcov2lcov.cpp:    for (i = 0; i < HASHTABLE_SIZE(line_htable.table_bits); i++) {
ext/drwrap/drwrap.c:        for (i = 0; i < HASHTABLE_SIZE(wrap_table.table_bits); i++) {
ext/drx/drx.c:    for (i = 0; i < HASHTABLE_SIZE(job_table.table_bits); i++) {
suite/tests/client-interface/custom_traces.dll.c:#define HASHTABLE_SIZE(num_bits) (1U << (num_bits))
suite/tests/client-interface/custom_traces.dll.c:#define TABLE_SIZE HASHTABLE_SIZE(HASH_BITS) * sizeof(trace_head_entry_t*)
suite/tests/client-interface/custom_traces.dll.c:    for (i = 0; i < HASHTABLE_SIZE(HASH_BITS); i++) {
suite/tests/client-interface/pcache.dll.c:    for (i = 0; i < HASHTABLE_SIZE(sample_pointer_table.table_bits); i++) {
suite/tests/client-interface/pcache.dll.c:    for (i = 0; i < HASHTABLE_SIZE(sample_inlined_table.table_bits); i++) {
```

There are also many uses in Dr. Memory:

```
$ git grep HASHTABLE_SIZE | grep -v winsyms
common/alloc.c:        for (i = 0; i < HASHTABLE_SIZE(alloc_routine_table.table_bits); i++) {
common/alloc.c:    for (i = 0; i < HASHTABLE_SIZE(malloc_table.table_bits); i++) {
common/alloc_replace.c:    for (i = 0; i < HASHTABLE_SIZE(pre_us_table.table_bits); i++) {
common/alloc_replace.c:            for (i = 0; i < HASHTABLE_SIZE(pre_us_table.table_bits); i++) {
common/alloc_replace.c:    for (i = 0; i < HASHTABLE_SIZE(pre_us_table.table_bits); i++) {
common/alloc_replace.c:    for (i = 0; i < HASHTABLE_SIZE(crtheap_handle_table.table_bits); i++) {
common/utils.c:    for (i = 0; i < HASHTABLE_SIZE(table->table_bits); i++) {
drheapstat/drheapstat.c:        for (i = 0; i < HASHTABLE_SIZE(alloc_stack_table.table_bits); i++) {
drmemory/gdicheck.c:    for (i = 0; i < HASHTABLE_SIZE(dc_table.table_bits); i++) {
drmemory/handlecheck.c:    for (i = 0; i < HASHTABLE_SIZE(table->table_bits); i++) {
drmemory/instru.c:    for (i = 0; i < HASHTABLE_SIZE(stringop_app2us_table.table_bits); i++) {
drmemory/report.c:    for (i = 0; i < HASHTABLE_SIZE(error_table.table_bits); i++) {
drsymcache/drsymcache.c:    for (i = 0; i < HASHTABLE_SIZE(symtable->table_bits); i++) {
drsymcache/drsymcache.c:    for (i = 0; i < HASHTABLE_SIZE(symcache_table.table_bits); i++) {
drsyscall/drsyscall.c:    for (i = 0; i < HASHTABLE_SIZE(systable.table_bits); i++) {
```

Naturally this CL would not update these Dr. Memory instances, but it would be useful to ensure the interface could be used there without changes.  Glancing at some of them, it may be awkward to have an interface where the iteration logic is in a separate function -- the alternative is a stateful iterator data structure though which is more complex to implement.  Should we eventually provide both types of iterators, or if we did create a synchronous iterator at some point would it completely replace this function call-based version?

Also, the interface should make it clear whether it supports calling hashtable_remove*() while iterating.

As with all user-visible features, please add a line to the changelist https://github.com/DynamoRIO/dynamorio/blob/master/api/docs/release.dox#L216

Finally, please add some sort of test to suite/tests/client-interface/drcontainers-test.dll.c.

Sorry to throw so much at you.

**Comment by johnfxgalea** [comment]
@derekbruening @fhahn  Thanks for the comments.

I agree that an iterator data structure is of use, and may be more suitable in some cases. With such functionality, it may appear that the function call approach is unneeded. The only benefit I see with the proposed approach, is that it avoids the developer to include iterator logic such as `start` and `next` (along with NULL checks on next within a loop statement).

**Comment by derekbruening** [comment]
It's up to you: if you are happy with your proposed interface here, it seems ok to add it now (ideally with a test added, plus @fhahn's code comments of course), and someone can file a feature request in the issue tracker for a synchronous iterator to be added in the future that then replaces most or all of the existing cases that do raw iteration (put an XXX comment xref to the new issue in the code).

**Comment by johnfxgalea** [comment]
Are there any tests already done for hashtables? The [drcontainers-test.dll.c](suite/tests/client-interface/drcontainers-test.dll.c) file seems to only contain tests for drvector.

**Comment by AssadHashmi** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
There is a trailing space in one of your commits, that causes the failure, e.g. https://travis-ci.org/DynamoRIO/dynamorio/jobs/318650290#L772

**Comment by AssadHashmi** [comment]
run aarch64 tests

**Comment by derekbruening** [comment]
The make/git/devsetup.sh script (see https://github.com/DynamoRIO/dynamorio/wiki/Workflow) blocks trailing spaces pre-commit.

**Comment by johnfxgalea** [comment]
@derekbruening It's unclear to me what I need to do next. Thanks!

**Comment by derekbruening** [comment]
> Are there any tests already done for hashtables? The drcontainers-test.dll.c file seems to only contain tests for drvector.

Unfortunately not: in the distant past testing for some components was neglected, but we would like to avoid adding new features without tests.  Just a simple test that at least exercises the function should be straightforward to add.

I filed #2782 so I suppose this is a partial implementation of that.

File: api/docs/release.dox [hunk 3]
```diff
@@ -212,6 +212,8 @@ Further non-compatibility-affecting changes include:
    different types of entries in a trace, broken down by thread.
  - Added instr_is_string_op() and instr_is_rep_string_op().
  - Added dr_app_recurlock_lock().
+ - Added hashtable_apply_to_all_payloads to iterate over all payloads in a
```
Comment (derekbruening) [inline api/docs/release.dox]: As you can see on the other lines, () is needed to get a doxygen link.

File: ext/drcontainers/hashtable.c [hunk 4]
```diff
@@ -457,6 +457,22 @@ hashtable_remove_range(hashtable_t *table, void *start, void *end)
     return res;
 }
 
+
+void
+hashtable_apply_to_all_payloads(hashtable_t *table, void (*apply_func)(void *payload))
+{
+    DR_ASSERT(apply_func);
```
Comment (derekbruening) [inline ext/drcontainers/hashtable.c]: style: explicit bools: != NULL

File: ext/drcontainers/hashtable.h [hunk 5]
```diff
@@ -193,6 +193,15 @@ hashtable_remove(hashtable_t *table, void *key);
 bool
 hashtable_remove_range(hashtable_t *table, void *start, void *end);
 
+/**
+ * Calls the apply_func for each payload
+ *
+ * @param apply_func A pointer to a function that is called for all payloads in
```
Comment (derekbruening) [inline ext/drcontainers/hashtable.h]: s/in//

File: ext/drcontainers/hashtable.h [hunk 6]
```diff
@@ -193,6 +193,15 @@ hashtable_remove(hashtable_t *table, void *key);
 bool
 hashtable_remove_range(hashtable_t *table, void *start, void *end);
 
+/**
+ * Calls the apply_func for each payload
```
Comment (derekbruening) [inline ext/drcontainers/hashtable.h]: nit: `\p apply_func`
nit: `.` at end

File: ext/drcontainers/hashtable.c (hunk 4 above)
Comment (fhahn) [reply ext/drcontainers/hashtable.c]: Could you also use DR_ASSERT_MSG and provide a message when this assertion is hit?

File: ext/drcontainers/hashtable.h (hunk 5 above)
Comment (johnfxgalea) [reply ext/drcontainers/hashtable.h]: Thanks, will get these sorted.

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 7]
```diff
@@ -76,10 +77,35 @@ test_vector(void)
...
+
+static void
+test_hashtable_apply_all(void){
+
+    hashtable_t hash_table;
+    hashtable_init(&hash_table, 8, HASH_INTPTR, false);
+    
+    c = 0;
+
+    hashtable_add_replace(&hash_table, (void *) 1, (void *)  1);
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: `(void *)  1`

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 8]
```diff
@@ -76,10 +77,35 @@ test_vector(void)
     CHECK(ok, "drvector_delete failed");
 }
 
+unsigned int c;
+
+void count(void *payload){
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: Type and { on separate lines, as in style guide (https://github.com/DynamoRIO/dynamorio/wiki/Code-Style)

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 9]
```diff
@@ -76,10 +77,35 @@ test_vector(void)
...
+    c++;
+}
+
+static void
+test_hashtable_apply_all(void){
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: { on newline

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 10]
```diff
@@ -76,10 +77,35 @@ test_vector(void)
...
+
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: no space here

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 11]
```diff
@@ -76,10 +77,35 @@ test_vector(void)
     CHECK(ok, "drvector_delete failed");
 }
 
+unsigned int c;
+
+void count(void *payload){
+    c++;
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: Could you also add a test that accesses the payload, e.g. sum()?

**Comment by fhahn** [comment]
For some reason I still had to resolve the conflict via the Github interface.

run aarch64 tests

**Comment by fhahn** [comment]
run aarch64 tests

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 12]
```diff
@@ -76,10 +77,49 @@ test_vector(void)
...
+
+static void 
+count(void *payload)
+{
+    c++;
+}
+
+unsigned int total;
+
+static void
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: Trailing space here. This trips over the bots

**Comment by fhahn** [comment]
Looks like there still are trailing whitespaces.

**Comment by fhahn** [comment]
Doxygen failed, but I could not find the exact error message from the Travis logs.

File: ext/drcontainers/hashtable.h [hunk 13]
```diff
@@ -193,6 +193,14 @@ hashtable_remove(hashtable_t *table, void *key);
 bool
 hashtable_remove_range(hashtable_t *table, void *start, void *end);
 
+/**
+ * Calls the \p apply_func for each payload.
+ * @param apply_func A pointer to a function that is called for all payloads in
```
Comment (fhahn) [inline ext/drcontainers/hashtable.h]: The table parameter needs a doxygen comment too, sorry for not spotting this

```
ext/drcontainers/hashtable.h:202:
  warning: The following parameters of
  hashtable_apply_to_all_payloads(hashtable_t *table, void(*apply_func)(void
  *payload)) are not documented:

    parameter 'table'

```

**Comment by fhahn** [comment]
run aarch64 tests

File: ext/drcontainers/hashtable.h (hunk 13 above)
Comment (derekbruening) [inline ext/drcontainers/hashtable.h]: grammar: s/in//

**Comment by fhahn** [comment]
Travis is failing because of a cast issue: 	`/.../dynamorio/suite/tests/client-interface/drcontainers-test.dll.c:93:12: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]`

**Comment by johnfxgalea** [comment]
A double case ```(unsigned int) (uintptr_t)``` should do it or do you recommend with respect to DynamoRIO coding convention.

File: suite/tests/client-interface/drcontainers-test.dll.c [hunk 14]
```diff
@@ -90,7 +91,7 @@ unsigned int total;
 static void
 sum(void *payload)
 {
-  total += (unsigned int) payload;
+  total += (uintptr_t) payload;
```
Comment (fhahn) [inline suite/tests/client-interface/drcontainers-test.dll.c]: You might have to change the type of total to uintptr_t too

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
Committed with tweaked commit message. Thanks for baring with me

**Comment by johnfxgalea** [comment]
No worries! Thanks @fhahn and @derekbruening . Sorry about the trailing white-spaces. Definitely need to set up my editor to remove them!

**Pull request merged**